kernel/arch/dreamcast/hardware/spu.c

```diff
@@ -60,30 +56,28 @@ void spu_memload(uint32 dst, void *src_void, int length) {
 
     if(length > 0) {
         g2_fifo_wait();
-        g2_write_block_32((uint32*)src, dst, length);
+        g2_write_block_32((uint32_t *)src, dst, length);
     }
 }
 
-void spu_memload_sq(uint32 dst, void *src_void, int length) {
-    uint8 *src = (uint8 *)src_void;
+void spu_memload_sq(uintptr_t dst, void *src_void, size_t length) {
+    uint8_t *src = (uint8_t *)src_void;
     int aligned_len, old;
 
-    /* Make sure it's an even number of 32-bit words and convert the
-       count to a 32-bit word count */
+    /* Round up to the nearest multiple of 4 */
     if(length & 3) {
         length = (length + 4) & ~3;
     }
 
-    /* Add in the SPU RAM base (cached area) */
-    dst += 0x00800000;
-
     /* Using SQs for all that is divisible by 32 */
     aligned_len = length & ~31;
     length &= 31;
 
+    /* Add in the SPU RAM base (cached area) */
+    dst += SPU_RAM_BASE;
+
     old = irq_disable();
-    do { } while(*(vuint32 *)0xa05f688c & (1 << 5)); /* FIFO_SH4 */
-    do { } while(*(vuint32 *)0xa05f688c & (1 << 4)); /* FIFO_G2 */
```

Comment:
It must be replaced to g2_lock

Reply:
The new g2_lock function doesnt wait on FIFO_G2:

`static inline g2_ctx_t g2_lock(void) {
    g2_ctx_t ctx;

    ctx.irq_state = irq_disable();

    /* Suspend any G2 DMA */
    G2_DMA_SUSPEND_SPU = 1;
    G2_DMA_SUSPEND_BBA = 1;
    G2_DMA_SUSPEND_CH2 = 1;

    while(FIFO_STATUS & FIFO_SH4);

    return ctx;
}` Should that be updated too?

Reply:
Just add g2_fifo_wait() after g2_lock()

Reply:
In general, there are different cases. Somewhere we only need to wait for SH4, somewhere SH4 + G2, and somewhere SH4 + G2 + AICA.
In this case we need to wait SH4 + G2. In case with AICA registers, need wait AICA FIFO in additional. 
And also every 8 regs writes need wait AICA FIFO, as 8 writes to memory we wait G2 FIFO.

Reply:
DONE